AMP atoms

nicl opened the pull request:
Note: requires https://github.com/guardian/frontend/pull/21357 to actually bear fruit (although can be merged before this).

## What does this change?

Adds support for a few atom types:

* guide
* profile
* q+a
* timeline

## Why?

AMP completion!

## Screenshots

<img width="606" alt="Screenshot 2019-04-23 at 11 31 14" src="https://user-images.githubusercontent.com/858402/56599270-217d6a80-65ee-11e9-92ec-40b49d2a3cf4.png">
<img width="606" alt="Screenshot 2019-04-23 at 12 25 35" src="https://user-images.githubusercontent.com/858402/56599271-22160100-65ee-11e9-856a-1d65f0a78dfb.png">
<img width="606" alt="Screenshot 2019-04-23 at 17 07 32" src="https://user-images.githubusercontent.com/858402/56599280-27734b80-65ee-11e9-890e-a1980a8f9418.png">

Note, again I've kind of adopted mostly mobile styling here as the AMP styling is a bit wacky/out of date - especially for the timeline atom.

PRBuilds commented:
PRbuilds results:














**LightHouse Reporting**




--automated message

nicl commented on packages/frontend/amp/components/Expandable.tsx:
```diff
@@ -0,0 +1,180 @@
+        )
+        13;
+
+    a {
```
Adding styling for `a`, `p`, `b` etc feels wrong here - I should grab this from elsewhere as it's general styling for text markup.

nicl replied:
yeah I agree. So I am about to add a commit to export the styling from the `Text` element component and use that. But really we probably need to think about this. My own view is that it should live in Guui but I know at the moment that has been purely for actual React components.

vlbee commented on packages/frontend/amp/components/elements/Timeline.tsx:
```diff
@@ -0,0 +1,82 @@
+`;
+
+const getHTML = (events: TimelineEvent[], description?: string): string => {
+    // TODO using ternary as && doesn't seem to work here - it prints out undefined
```
Is this TODO still supposed to be here? (don't see a ternary)

nicl replied:
Oh yeah thanks. I am a bit confused here about how JSX handles && when the first argument is undefined but perhaps it is now okay? Will check/remove comment but if anyone knows what to check here let me know!

gtrufitt replied:
Did you resolve this>

gtrufitt commented on packages/frontend/amp/components/Expandable.tsx:
```diff
@@ -0,0 +1,180 @@
+            <span className={cx(headerStyle, pillarColour(pillar))}>
+                {type}
+            </span>
+            <h1 className={headerStyle}>{title}</h1>
```
Pinging @SiAdcock on this one - I know this is in an aside, but I'm not keen on using `h1`s within the content in this way because there can be problems with assistive technology. *However*, my knowledge could be out of date so I'd defer to @SiAdcock on his thoughts.

nicl replied:
okay thanks, yeah I have no idea here.

nicl replied:
For the record, we've gone with h2 here, and then h3's for contained headings.

nicl commented on packages/frontend/amp/components/elements/Timeline.tsx:
```diff
@@ -0,0 +1,82 @@
+    ));
+
+    const eventString = renderToStaticMarkup(
+        <ul className={eventsWrapper}>{eventMarkup}</ul>,
```
yeah this is forced by prettier, but it doesn't affect the markup.

nicl commented on packages/frontend/amp/components/Expandable.tsx:
```diff
@@ -0,0 +1,180 @@
+    padding-bottom: 24px;
+`;
+
+const iconStyle = css`
```
Ah cool, wasn't aware Support-UI was far enough along to use, but that's great.

gtrufitt replied:
I think it'd be ideal to bring the button stuff into GUUI

nicl commented on packages/frontend/amp/components/elements/Timeline.tsx:
```diff
@@ -0,0 +1,82 @@
+    // TODO using ternary as && doesn't seem to work here - it prints out undefined
+    const eventMarkup = events.map(e => (
+        <li className={eventStyle}>
+            <time className={eventIconStyle}>{e.date}</time>
```
Yeah - at the moment I've just passed them as strings to avoid duplicating the formatting logic from Frontend, but I'd like to pass through an actual date so that we can do this and format as appropriate in DCR.

We can definitely do this now though so shout if you think it's worthwhile. (At the moment I've got a TODO in Frontend around the Timeline case class noting that we should really pass through a richer date object.)

gtrufitt replied:
I think there's probably a wider questions around machine readable data. I couldn't find a specific `timeline` type reference in Schema but something like https://schema.org/EventSeries might work. I don't know if it's worthwhile now or looking at holistically across everything a bit closer - `datetime` probably doesn't give you a whole lot of advantage alone in this specific case.

nicl replied:
ah yes definitely. There is probably stuff to look at after we get live here - which I realise means it may never happen - but I'd prefer product to appreciate the cost/prioritise here than mask it in a slower migration.

nicl commented on packages/frontend/amp/components/elements/Timeline.tsx:
```diff
@@ -0,0 +1,82 @@
+            <time className={eventIconStyle}>{e.date}</time>
+            {e.toDate && (
+                <>
+                    {' '}
```
:cry:

nicl commented on packages/frontend/amp/components/Expandable.tsx:
```diff
@@ -0,0 +1,180 @@
+                <amp-img
+                    class={imageStyle}
+                    src={img}
+                    alt={`Image for ${title} explainer`}
```
Will check what we get on the models, but at the moment AMP just uses the type of the atom here - e.g. 'Guide' which doesn't seem especially useful!

nicl commented:
@gtrufitt I think it's okay. Originally I added the comment when I was doing string interpolation rather than JSX elements for this (long story) which I think works differently. I've removed the comment after testing to make sure!
